### PR TITLE
Restore formal grammar

### PR DESCRIPTION
**Pull request by sjamaan**
As per #24 I would really like to see a formal grammar in the URL spec.

This is an initial start on such a grammar. I've taken the ABNF from RFC3986.  As far as I can tell, the WHATWG URL spec has removed support for the `ipvfuture` production, so I've ripped that out.  I've also **not** added support for IPv6 zone identifiers from RFC6874, as the WHATWG spec explicitly states

> Support for `<zone_id>` is intentionally omitted.

I'm not sure how to best proceed to add error handling. I think it's probably best to first ensure that the syntax described by the ABNF matches the syntax accepted by the algorithm (without any errors or warnings). Where the algorithm accepts a URL that is not supported by the ABNF, the ABNF could be updated if it the algorithm would not raise a validation error. If there are URLs that will be accepted by the ABNF, the algorithm probably needs to be updated, unless it that syntax is intentionally no longer supported.

***
<a href="https://whatpr.org/url/416.html" title="Last updated on Jan 15, 2021, 7:42 AM UTC (ccadabf)">Preview</a> | <a href="https://whatpr.org/url/416/85d96e2...ccadabf.html" title="Last updated on Jan 15, 2021, 7:42 AM UTC (ccadabf)">Diff</a>

### CODE REVIEW

File: url.bs [hunk 1]
```diff
@@ -2346,6 +2347,105 @@ string <var>input</var>, optionally with a <a>base URL</a> <var>base</var>, opti
...
 
+<h4 id=grammar>Formal grammar for minimally acceptable URLs</h4>
+
+<p>It is the intention that the above parsing algorithm accepts a
+superset of the language defined by the following grammar.  The
+grammar is only defined for UTF-8.  The <var>encoding override</var>
+option is not supported.  Resolving the URL against a <a>base URL</a>
+or raising an error if the URL is relative should be done after
+parsing.</p>
+
```
Comment (GPHemsley) [inline url.bs]: All text within the specification is to be considered as normative. However, this sounds more like a note to me than a normative requirement.

File: url.bs [hunk 2]
```diff
@@ -2346,6 +2347,105 @@ string <var>input</var>, optionally with a <a>base URL</a> <var>base</var>, opti
...
+<p>This specification uses the Augmented Backus-Naur Form (ABNF)
+notation of STD 68 [[!STD68]], including the following core ABNF
+syntax rules defined by that specification: ALPHA (letters), CR
+(carriage return), DIGIT (decimal digits), DQUOTE (double quote),
+HEXDIG (hexadecimal digits), LF (line feed), and SP (space).</p>
+
```
Comment (GPHemsley) [inline url.bs]: I think it's generally preferred that terms defined in other standards are linked to their definitions in those other standards.

File: url.bs [hunk 3]
```diff
@@ -2346,6 +2347,105 @@ string <var>input</var>, optionally with a <a>base URL</a> <var>base</var>, opti
...
+   pct-encoded   = "%" HEXDIG HEXDIG
+
+   unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
+   reserved      = gen-delims / sub-delims
+   gen-delims    = ":" / "/" / "?" / "#" / "[" / "]" / "@"
+   sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
+                 / "*" / "+" / "," / ";" / "="
+</pre>
+
```
Comment (GPHemsley) [inline url.bs]: I think more needs to be done here to make this useful than to just drop a huge chunk of preformatted text into the spec. Also, I think wpt would need to be updated to somehow test this grammar as distinct from the parser algorithm.

File: url.bs (hunk 1 above)
Comment (sjamaan) [reply url.bs]: How should I reword it to make it normative?

File: url.bs (hunk 3 above)
Comment (sjamaan) [reply url.bs]: Regarding the preformatted text, I wouldn't mind polishing it to make it formatted in a way that makes it more readable. Would that involve adding BNF tags to bikeshed, or are there suitable options already?

Testing a grammar would be kind of tricky. Usually one would pick a parser generator and make the necessary transformations to make the BNF fit that particular parser. I have several different implementations of the RFC using various parser generators in [my project's repository](https://bugs.call-cc.org/browser/project/release/5/uri-generic/trunk/alternatives) but if you inspect them they all involve some modifications. Funny enough, the [one based on SRE](https://bugs.call-cc.org/browser/project/release/5/uri-generic/trunk/alternatives/uri-generic.irregex.scm#L503) (a composable form of regular expressions) is the only one without *any* changes from the BNF, IIRC.

We could include that SRE-based implementation in WPT. If only JavaScript is accepted as an implementation language I can take a look at the ecosystem to find a good parser generator and try my hand at transcoding the BNF to it. I'm not a good JavaScript coder, though!

File: url.bs (hunk 2 above)
Comment (sjamaan) [reply url.bs]: I copied this from the RFC. I did link to the ABNF STD. How would you recommend I change this?

**Comment by sjamaan** [comment]
hm, I just clicked "resolve conversation". I'm not sure that's the correct usage of that button. I thought that would actually submit the comments but it looks like it just closes the "subthread". Sorry about that!
